drivers/dht/dht.c

```diff
@@ -19,192 +20,267 @@
  * @author      Ludwig Knüpfer <ludwig.knuepfer@fu-berlin.de>
  * @author      Christian Mehlis <mehlis@inf.fu-berlin.de>
  * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @author      Hugues Larrive <hugues.larrive@pm.me>
  *
  * @}
  */
 
+#include <errno.h>
 #include <stdint.h>
 #include <string.h>
 
-#include "log.h"
 #include "assert.h"
-#include "xtimer.h"
-#include "timex.h"
-#include "periph/gpio.h"
-
 #include "dht.h"
 #include "dht_params.h"
+#include "log.h"
+#include "periph/gpio.h"
+#include "time_units.h"
+#include "ztimer.h"
 
 #define ENABLE_DEBUG            0
 #include "debug.h"
 
 /* Every pulse send by the DHT longer than 40µs is interpreted as 1 */
-#define PULSE_WIDTH_THRESHOLD   (40U)
-/* If an expected pulse is not detected within 1000µs, something is wrong */
-#define TIMEOUT                 (1000U)
-/* The DHT sensor cannot measure more than once a second */
-#define DATA_HOLD_TIME          (US_PER_SEC)
-/* The start signal by pulling data low for at least 18ms and then up for
- * 20-40µs*/
-#define START_LOW_TIME          (20U * US_PER_MS)
-#define START_HIGH_TIME         (40U)
-
-static inline void _reset(dht_t *dev)
+#define READ_THRESHOLD          (40U)
+/* If an expected pulse is not detected within 85µs, something is wrong */
+#define SPIN_TIMEOUT            (85U)
+/* According to datasheet the DHT sensor cannot measure more than once
+ * every 2 seconds but most are faster */
+#define RESPAWN_TIMEOUT         (2U * US_PER_SEC)
+/* The start signal by pulling data low for at least 18 ms for DHT11, at
+ * most 20 ms (AM2301 / DHT22 / DHT21). Then release the bus and the
+ * sensor should respond by pulling data low for 80 µs, then release for
+ * 80µs before start sending data. */
+#define START_LOW_TIME          (19U * US_PER_MS)
+#define START_THRESHOLD         (75U)
+/* DHTs have to wait for power 1 or 2 seconds depending on the model */
+#define POWER_WAIT_TIMEOUT      (2U * US_PER_SEC)
+
+enum {
+    BYTEPOS_HUMIDITY_HIGH = 0,
+    BYTEPOS_HUMIDITY_LOW = 1,
+    BYTEPOS_TEMPERATURE_HIGH = 2,
+    BYTEPOS_TEMPERATURE_LOW = 3,
+    BYTEPOS_CHECKSUM = 4,
+};
+
+struct dht_data {
+    gpio_t pin;
+    gpio_mode_t in_mode;
+    uint8_t data[5];
+    int8_t bit_pos;
+    uint8_t bit;
+};
+
+static inline void _wait_for_level(gpio_t pin, bool expected, uint32_t start)
 {
+    uint8_t pre_timeout = 0;
+    while ((gpio_read(pin) != expected) && (++pre_timeout
+        || ztimer_now(ZTIMER_USEC) < start + SPIN_TIMEOUT)) {}
+}
+
+static inline int8_t _send_start_signal(dht_t *dev)
```

Comment:
```suggestion
static inline int _send_start_signal(dht_t *dev)
```

`errno` codes are allowed to make use of the whole `int` range. E.g. newlib has

```C
#define ESTALE 133
#define ENOTSUP 134		/* Not supported */
/* ... */
#define __ELASTERROR 2000	/* Users can add values starting here */
```

So at least some are already beyond the `int8_t` range. Technically, `uint8_t` would be sufficient as of now, but looking at `__ELASTERROR`, we would need `int_least16_t` or just `int` to be safe here. Given that `errno` is defined as `int`, I'd prefer `int` here.